### PR TITLE
Keep variables for constant zero Polynomials

### PR DESCRIPTION
**Pull request by manuelbb-upb**
Fix of [issue #80](https://github.com/JuliaAlgebra/DynamicPolynomials.jl/issues/80)

Overwrite `multconstant` of *MultivariatePolynomials* so that in case of `α==0` the variables of `p` are kept.
Else do the same as before (see `operators.jl` in *MultivariatePolynomials*).

### CODE REVIEW

**Comment by blegat** [comment]
Does this fix https://github.com/JuliaAlgebra/DynamicPolynomials.jl/issues/79 as well ?

**Comment by manuelbb-upb** [comment]
> Does this fix #79 as well ?

No, it doesn't seem to.
But I can have a look into it in the near future.

**Comment by manuelbb-upb** [comment]
It seems that #79 can be tackled by defining
```
function MA.mutable_operate_to!(p::Polynomial{true, T}, ::typeof(*), q1::Number, q2::MP.AbstractPolynomialLike) where T
    q1 * q2
end
```
This uses
```
function Base.:(*)(p::Polynomial, α)
    MP.multconstant(α,p);
end
Base.:(*)(α, p::Polynomial) = p * α;
```
but I am not familiar enough with the internals of the package to know whether the fix is any good.

File: src/mult.jl [hunk 1]
```diff
@@ -29,6 +29,17 @@ include("cmult.jl")
 include("ncmult.jl")
 
 MP.multconstant(α, x::Monomial)   = MP.term(α, MA.mutable_copy(x))
+
+function MP.multconstant(α, p :: Polynomial )
```
Comment (blegat) [inline src/mult.jl:33]: I would define a function `zero_with_variables` and do
```julia
function _multconstant(α::T, f, p::AbstractPolynomial{S}) where {S, T}
    if iszero(α)
        zero_with_variables(polynomialtype(p, MA.promote_operation(*, T, S)), variables(p))
    else
        mapcoefficientsnz(f, p)
    end
end
```

File: src/mult.jl [hunk 2]
```diff
@@ -60,7 +71,7 @@ end
 function _term_poly_mult(t::Term{C, S}, p::Polynomial{C, T}, op::Function) where {C, S, T}
     U = MA.promote_operation(op, S, T)
     if iszero(t)
-        zero(Polynomial{C, U})
+        zero(p)
```
Comment (blegat) [inline src/mult.jl:74]: Use `zero_with_variables(Polynomial{C, U}, variables(p))`

File: src/mult.jl (hunk 1 above)
Comment (manuelbb-upb) [reply src/mult.jl:33]: As suggested I have defined `zero_with_variables` as 
```
function zero_with_variables( ::Type{Polynomial{C,T}}, vars :: Vector{PolyVar{C}} ) where{C, T}
    Polynomial( T[], emptymonovec(vars) )
end
```
Overwriting `_multconstant` via 
```
function MP._multconstant(α::T, f, p::AbstractPolynomial{S}) where {S, T}
    if iszero(α)
        zero_with_variables(polynomialtype(p, MA.promote_operation(*, T, S)), variables(p))
    else
        mapcoefficientsnz(f, p)
    end
end
```
works but breaks something ( `** incremental compilation may be fatally broken for this module **` )

File: src/mult.jl (hunk 2 above)
Comment (manuelbb-upb) [reply src/mult.jl:74]: My first guess was wrong, this was not the culprit. I don't know whether to keep `zero(Polynomial{C, U})` or use `zero_with_variables( Polynomial{C,U}, variables(p) )`.

File: src/mult.jl (hunk 1 above)
Comment (blegat) [reply src/mult.jl:33]: It means you have defined twice the same method, remove the other one you have defined.

File: src/mult.jl (hunk 2 above)
Comment (blegat) [reply src/mult.jl:74]: You can leave as it is then. We can make the change in another PR.

File: src/mult.jl (hunk 1 above)
Comment (manuelbb-upb) [reply src/mult.jl:33]: I did change the function signature to apply to DynamicPolynomials  only, i.e., `MP._multconstant(α::T, f, p::Polynomial{C,S} ) where {T, C, S}` 

That does the trick

File: src/mult.jl [hunk 3]
```diff
@@ -138,3 +151,20 @@ end
...
 end
+
+# Overwrite this method for monomial-like terms because
+# otherwise it would check `iszero(α)` and in that case 
+# dismiss of the variable of `p` by performing 
+# `operate_to!(zero, output :: Polynomial )` which only
+# respects the variables that are stored already
+function MP._multconstant_to!(output::Polynomial, α, f, p :: DMonomialLike)
+    if iszero(α)
+        empty!(output.a)
```
Comment (blegat) [inline src/mult.jl:162]: Use `MA.mutable_operate!(zero, p)`

File: src/mult.jl [hunk 4]
```diff
@@ -138,3 +151,20 @@ end
...
+        empty!(output.x.vars)
+        push!(output.x.vars, variables(p)...)
```
Comment (blegat) [inline src/mult.jl:164]: Use https://github.com/JuliaAlgebra/DynamicPolynomials.jl/blob/9e4d3249d4b3c87df7179c213a67e7750c6e8217/src/monovec.jl#L255
`variables(p)...` will have to compile a new method for each number of variables

File: src/mult.jl [hunk 5]
```diff
@@ -60,7 +73,7 @@ end
 function _term_poly_mult(t::Term{C, S}, p::Polynomial{C, T}, op::Function) where {C, S, T}
     U = MA.promote_operation(op, S, T)
     if iszero(t)
-        zero(Polynomial{C, U})
+        zero( Polynomial{C,U} )
```
Comment (blegat) [inline src/mult.jl:76]: Remove spaces before and after parenthesis to stay consistent with the rest of the code base

File: src/mult.jl [hunk 6]
```diff
@@ -29,6 +29,19 @@ include("cmult.jl")
 include("ncmult.jl")
 
 MP.multconstant(α, x::Monomial)   = MP.term(α, MA.mutable_copy(x))
+
+function zero_with_variables( ::Type{Polynomial{C,T}}, vars :: Vector{PolyVar{C}} ) where{C, T}
+    Polynomial( T[], emptymonovec(vars) )
```
Comment (blegat) [inline src/mult.jl:34]: Remove spaces before and after parenthesis to stay consistent with the rest of the code base

**Comment by blegat** [comment]
Thanks, looking good, could you add some tests ?

**Comment by manuelbb-upb** [comment]
I have composed some basic tests from the issue descriptions. The nightly built still fails, but this seems to a separate problem.

**Review by blegat — APPROVED** [review]
Awesome, thanks a lot! Could your remove the Draft mode of the PR so that I can merge ?

**Comment by manuelbb-upb** [comment]
Did so :) Thanks also for reviewing. This was the first time for me to actually try to contribute to another project and a great learning experience.

**Pull request merged**